gammapy/modeling/models/spatial.py

```diff
@@ -109,9 +121,12 @@ def position_error(self):
         )
 
     def evaluate_geom(self, geom):
-        """Evaluate model on `~gammapy.maps.Geom`."""
         coords = geom.get_coord(frame=self.frame)
-        return self(coords.lon, coords.lat)
+        energy_name = _get_energy_axis(geom)
+        if energy_name is not None:
+            return self(coords.lon, coords.lat, coords[energy_name])
```

Comment:
I think we should rely on numpy broadcasting here, just as we do in `SkyModel.evaluate_geom()`, so evaluate the model not using the full 3D coordinate arrays, but relying on the outer product between the spatial and energy coordinates...

Reply:
Is this working for a full 3d template model ? SkyDiffuseCube use the SkyModelBase.evaluate_geom that look like this one, no ?

Reply:
Yes, it should work. In `Map.interp_by_coord()`, we call `MapCoord.create()`, which does the broadcasting of the arrays on init:

```python3
from gammapy.maps import MapAxis, Map
import numpy as np

axis = MapAxis.from_energy_bounds("1 TeV", "10 TeV", nbin=10)

m = Map.create(width=10, axes=[axis])

m.interp_by_coord(
{
    "skycoord": m.geom.to_image().get_coord().skycoord,
    "energy": axis.center.reshape((-1, 1, 1))
}
)
```